src/exporters/staking.ts

```diff
@@ -0,0 +1,233 @@
+		let ledgers = await api.query.staking.ledger.entries();
+	
+		let totalBondedAccounts = ledgers.length;
+		let totalBondedStake = ledgers.map(([_, ledger]) =>
+			ledger.unwrapOrDefault().total.toBn().div(decimals(api)).toNumber()
+		).reduce((prev, next) => prev + next)
+		let totalUnbondingChunks = ledgers.map(([_, ledger]) => {
+			if (ledger.unwrapOrDefault().unlocking.length) {
+				return ledger.unwrapOrDefault().unlocking.map((unlocking) =>
+					unlocking.value.toBn().div(decimals(api)).toNumber()
+				).reduce((prev, next) => prev + next)
+			} else {
+				return 0
+			}
+		}
+		).reduce((prev, next) => prev + next)
+	
+		this.ledgerMetric.set({ type: "bonded_accounts", chain: chainName }, totalBondedAccounts)
+		this.ledgerMetric.set({ type: "bonded_stake", chain: chainName }, totalBondedStake)
+		this.ledgerMetric.set({ type: "unbonding_stake", chain: chainName }, totalUnbondingChunks)
+
+	}
+
+	async perHour(api: ApiPromise, chainName: string) { 
+		logger.info(`starting hourly scrape at ${new Date().toISOString()}`)
+
+		let stakingPromise = this.stakingHourly(api, chainName);
+		let voterBagsPromise = this.voterBags(api, chainName);
+		Promise.all([stakingPromise, voterBagsPromise])
+
+	}
+	
+	async stakingHourly(baseApi: ApiPromise, chainName: string) {
+		const api = await getFinalizedApi(baseApi);
+		let currentEra = (await api.query.staking.currentEra()).unwrapOrDefault();
+		let exposures = await api.query.staking.erasStakers.entries(currentEra);
+	
+		let totalSelfStake = exposures.map(([_, expo]) => expo.own.toBn().div(decimals(api)).toNumber()).reduce((prev, next) => prev + next);
+		let totalOtherStake = exposures.map(([_, expo]) => (expo.total.toBn().sub(expo.own.toBn())).div(decimals(api)).toNumber()).reduce((prev, next) => prev + next);
+		let totalStake = totalOtherStake + totalSelfStake;
+	
+		this.stakeMetric.set({ type: "self", chain: chainName }, totalSelfStake);
+		this.stakeMetric.set({ type: "other", chain: chainName }, totalOtherStake);
+		this.stakeMetric.set({ type: "all", chain: chainName }, totalStake);
+	
+		let totalExposedNominators = exposures.map(([_, expo]) => expo.others.length).reduce((prev, next) => prev + next);
+		let totalExposedValidators = exposures.length;
+	
+		this.validatorCountMetric.set({ type: "active", chain: chainName }, totalExposedValidators);
+		this.nominatorCountMetric.set({ type: "active", chain: chainName }, totalExposedNominators);
+	}
+
+	async  voterBags(baseApi: ApiPromise, chainName: string) {
+		const api = await getFinalizedApi(baseApi);
+	
+		interface Bag {
+			head: AccountId,
+			tail: AccountId,
+			upper: Balance,
+			nodes: AccountId[],
+		}
+	
+		async function needsRebag(
+			api: ApiDecoration<"promise">,
+			bagThresholds: BN[],
+			node: PalletBagsListListNode,
+		): Promise<boolean> {
+			const currentWeight = await correctWeightOf(node, api);
+			const canonicalUpper = bagThresholds.find((t) => t.gt(currentWeight)) || new BN("18446744073709551615");
```

Comment:
where does this number come from?

Reply:
I recall it is the hardcoded max bag and it is actually my code that's being moved around: 

https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/bin/node/runtime/src/voter_bags.rs#L234